Lib/test/test_httpservers.py

```diff
@@ -315,6 +332,76 @@ def test_head_via_send_error(self):
             self.assertEqual(b'', data)
 
 
+@unittest.skipIf(ssl is None, 'No ssl module')
+class BaseHTTPSServerTestCase(BaseTestCase):
+    def _data_file(*name):
+        return os.path.join(os.path.dirname(__file__), "certdata", *name)
+
+    CERTFILE = _data_file("keycert.pem")
+    ONLYCERT = _data_file("ssl_cert.pem")
+    ONLYKEY = _data_file("ssl_key.pem")
+    CERTFILE_PROTECTED = _data_file("keycert.passwd.pem")
+    ONLYKEY_PROTECTED = _data_file("ssl_key.passwd.pem")
+    KEY_PASSWORD = "somepass"
+    EMPTYCERT = _data_file("nullcert.pem")
+    BADCERT = _data_file("badcert.pem")
+
+    tls = (ONLYCERT, ONLYKEY, None) # values by default
+
+    class request_handler(NoLogRequestHandler, SimpleHTTPRequestHandler):
+        pass
+
+    def test_get(self):
+        response = self.request('/')
+        self.assertEqual(response.status, HTTPStatus.OK)
+
+    def request(self, uri, method='GET', body=None, headers={}):
+        self.connection = http.client.HTTPSConnection(
+            self.HOST,
+            self.PORT,
+            context=ssl._create_unverified_context()
+        )
+        self.connection.request(method, uri, body, headers)
+        return self.connection.getresponse()
+
+    def test_valid_certdata(self):
+        valid_certdata_examples = (
+            (self.CERTFILE, None, None),
+            (self.CERTFILE, self.CERTFILE, None),
+            (self.CERTFILE_PROTECTED, None, self.KEY_PASSWORD),
+            (self.ONLYCERT, self.ONLYKEY_PROTECTED, self.KEY_PASSWORD),
+        )
+        for data in valid_certdata_examples:
+            server = HTTPSServer(
+                ('localhost', 0),
+                BaseHTTPRequestHandler,
+                certfile=data[0],
+                keyfile=data[1],
+                password=data[2],
+            )
+            self.assertIsInstance(server, HTTPSServer)
+            server.server_close()
+
+    def test_invalid_certdata(self):
+        invalid_certdata_examples = (
+            (self.BADCERT, None, None),
+            (self.EMPTYCERT, None, None),
+            (self.ONLYCERT, None, None),
+            (self.ONLYKEY, None, None),
+            (self.ONLYKEY, self.ONLYCERT, None),
+            (self.CERTFILE_PROTECTED, None, "badpass"),
+        )
+        for data in invalid_certdata_examples:
+            with self.assertRaises(ssl.SSLError):
+                HTTPSServer(
+                    ('localhost', 0),
+                    self.request_handler,
+                    certfile=data[0],
+                    keyfile=data[1],
+                    password=data[2],
+                )
+
+
```

Comment:
Can we test that `ImportError` is also raised? for this you can do:

```py
with import_helper.isolated_modules():
    sys.modules['ssl'] = None
    address = ('localhost', 0)
    with self.assertRaises(ImportError):
        HTTPSServer(address, self.request_handler, certfile=self.ONLYCERT)
```

Reply:
Ah you are actually raising a RuntimeError. Sorry, I thought it was an ImportError. Check that the documentation reflects the correct error by the way, thanks in advance.

Reply:
i replace it to `RuntimeError` in the documentation

Reply:
@picnixz and if we added this function to the class it will not be executed because we already have next decorator installed:
``` python
@unittest.skipIf(ssl is None, "requires ssl")
class BaseHTTPSServerTestCase(BaseTestCase): ...
```

Reply:
I wasn't actually thinking of adding it to the class. I thought about adding it before the `BaseTestCase` class (namely as a function not a method)

Reply:
i was just confused by the use of `self`, maybe we should make a separate class, like this:
```python
class TestSSLDisabled(unittest.TestCase):
    def test_https_server_raises_runtime_error(self):
        with isolated_modules():
            sys.modules['ssl'] = None
            address = ('localhost', 0)
            with self.assertRaises(RuntimeError):
                HTTPSServer(address, DummyRequestHandler, certfile="")
```

i'm not sure that we can call `assertRaises` without `self`

Reply:
Yes, either here or somewhere else. What I wanted to say is that you needed an additional test. I wasn't writing a suggestion to apply, just how I thought the test could look like.
I would suggest keeping it in `BaseHTTPSServerTestCase` as we could then use a valid certfile (even if `ssl` is imported first, what if, in the future, we import it only later; then if `certfile` is itself invalid, maybe we'll catch another exception).

Reply:
This is what I got:
```python
class TestSSLDisabled(unittest.TestCase):
    def test_https_server_raises_runtime_error(self):
        with import_helper.isolated_modules():
            sys.modules['ssl'] = None
            certfile = certdata_file("keycert.pem")
            with self.assertRaises(RuntimeError):
                create_https_server(certfile)

```